src/vscode-utils.ts

```diff
@@ -7,3 +14,64 @@ export function getVscode(): typeof import("vscode") {
     );
   }
 }
+
+export function parseJsonc(jsonc: string) {
+  const errors: ParseError[] = [];
+  const result = jsoncParse(jsonc, errors, {
+    allowEmptyContent: true,
+    allowTrailingComma: true,
+  });
+  if (errors.length) {
+    throw new AggregateError(errors, 'failed to parse JSONC');
+  }
+  return result;
+}
+
+/**
+ * A helper class to read files from extension directories. Uses a `jsonc` parser to parse JSON.
+ */
+export class ExtensionFileReader {
+  constructor(private readonly vscode: typeof import('vscode'), private readonly decoder = new TextDecoder('utf-8')) {}
+
+  async readFile(base: Uri, path?: string): Promise<string> {
+    const uri = path ? this.vscode.Uri.joinPath(base, path) : base;
+    const bytes = await this.vscode.workspace.fs.readFile(uri);
+    const text = this.decoder.decode(bytes);
+    return text;
+  }
+
+  async readJson<T>(base: Uri, path: string): Promise<T> {
+    const text = await this.readFile(base, path);
+    const json = parseJsonc(text);
+    return json as T;
+  }
+
+  /**
+   * TextMate grammar/language files can be written in JSON, CSON, YAML or XML Plist formats.
+   *
+   * If the extension does not make it clear, we probe the first character to attempt to parse it correctly.
+   *
+   * based on https://github.com/shikijs/textmate-grammars-themes/blob/main/scripts/shared/parse.ts
```

Comment:
It's a bummer it's not exported from @shikijs/vscode-textmate, which we already have.
Do you know about extensions that define their tmGrammer in cson, or plist? doesn't VSCode demand extensions to format it as JSON?

Reply:
update: I see they're using this for parsing:
https://github.com/microsoft/vscode-textmate/blob/76ab07aecfbd7e959ee4b55de3976f7a3ee95f38/src/parseRawGrammar.ts#L10

Can we use it too?
I guess that shiki decides to parse CSON and YAML files too but VSCode extensions are not going to have extensions with grammar in these formats

Reply:
I looked into it, but it does exactly what we do, but their plist parser is part of the code base, we use `fast-plist`, which is the same code but extracted to a separate library. So the end result is the same. We can drop `jsonc`, because vscode uses a plain `JSON.parse`.